resources/assets/js/admin/components/taxonomies/TaxonomyTermEditorTagAutocomplete.vue

```
@@ -0,0 +1,108 @@
+<template>
+	<div>
+		<div v-if="selected" class="autocomplete-selected">
+			{{selected.name}}
+			<span class="icon is-small clickable" @click="onSelect(null)"><i class="fa fa-close" aria-hidden="true"></i></span>
+		</div>
+		<div class="control" v-else>
+			<input class="input" v-model="search" placeholder="Wpisz nazwę tagu, który chcesz dołączyć lub utworzyć" />
+			<wnl-autocomplete
+				:items="autocompleteTags"
+				:onItemChosen="onSelect"
+				:isDown="true"
+			>
+				<template slot-scope="slotProps">
+					<div>
+						{{slotProps.item.name}}
+					</div>
+				</template>
+				<template slot="footer" v-if="autocompleteTags.length === 0 && search !== ''">
+					<div>
+						<div class="margin">
+							Nie mamy taga <strong>{{search}}</strong>
+						</div>
+						<div class="autocomplete-footer-button-container">
+							<button class="button" @click="onTagAdd">
+								<span class="icon is-small">
+									<i class="fa fa-plus" aria-hidden="true"></i>
+								</span>
+								<span>Dodaj nowy tag</span>
+							</button>
+						</div>
+					</div>
+				</template>
+			</wnl-autocomplete>
+		</div>
+	</div>
+</template>
+
+<style lang="sass" rel="stylesheet/sass" scoped>
+	@import 'resources/assets/sass/variables'
+
+	.autocomplete-selected
+		display: flex
+		justify-content: space-between
+		padding: $margin-small-minus
+
+	.autocomplete-footer-button-container
+		border-top: $border-light-gray
+		padding: $margin-base
+		text-align: right
+
+</style>
+
+<script>
+import {mapState, mapActions} from 'vuex';
+import {uniqBy} from 'lodash';
+
+import WnlAutocomplete from 'js/components/global/Autocomplete';
+
+export default {
+	props: {
+		selected: {
+			type: Object,
+			default: null,
```

can the default be object? I know you would have to change `v-if=selected` to `v-if=selected.name`.

I'm not pushing on it, just a preference to keep typing consistent and avoid using nulls if not forced to.

---

We can change it, but let's agree which way we should go with the team
@gitkuba @rogatty do you have any preferences on an empty object vs null to mark "no value" state?

---

You didn't ask me but I'll add my 5 cents.

Why objects not nulls:
- we keep the typing consistent. If a variable is treated as an object then declaring null somewhere as default may cause errors. It may be a case if we pass this variable down through many components

In my opinion the only thing that null gives us is easier checking for empty value, so any usages in if statements are much simpler.

The question is if a shortcut right now is worth possible error in the future.

---

Thanks for the clarification - it'll help guys to make a more concious judgement on this topic. I agree with everything. On the other hand, I'm worried about cases when the empty object is a valid value and should be threaded as "regular value" - not "no value".

---

I'd initialize with empty object if i had an intention to fill that object with some values later on. In case of properties passed from outside we expect to receive something in a predictable form. My opinion is `null` to explicitly tell we didn't receive, what we expected or no default at all and throw an error if calling statment doesn't meet expectations.

---

BTW I don't feel attached to that judgement, just trying to provide another point of view 🙂

---

For throwing an error, you can use `required` attribute. 

In this case, we use `null` to tell application there is no value, but it can change on user interaction to a proper object.

---

I don't think having `null` is a shortcut, I'd say that an empty object is a shortcut. `null` clearly means there is no tag selected and we should make sure that the app handles it. By having an empty object we don't need to handle it in other places as long as there is only one level of properties, but if there are more, it'll be harder to debug. I prefer failing fast.

---

To follow up on what @gitkuba said - we're not filling an object with values later on. We're either replacing an empty object or null with a new object. In this light, the empty object makes even less sense.